test/proxy_test.go

```diff
@@ -0,0 +1,666 @@
+ * limitations under the License.
+ *
+ */
+
+package test
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"testing"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/internal/resolver/delegatingresolver"
+	"google.golang.org/grpc/internal/stubserver"
+	"google.golang.org/grpc/internal/testutils"
+	testgrpc "google.golang.org/grpc/interop/grpc_testing"
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/resolver/manual"
+)
+
+const (
+	unresolvedTargetURI = "example.com"
+	unresolvedProxyURI  = "proxyExample.com"
+)
+
+// overwriteAndRestore temporarily replaces `HTTPSProxyFromEnvironment` with a
+// custom function and returns a function to restore the original.
+func overwriteAndRestore(customFunc func(req *http.Request) (*url.URL, error)) func() {
+	originalFunc := delegatingresolver.HTTPSProxyFromEnvironment
+	delegatingresolver.HTTPSProxyFromEnvironment = customFunc
+	return func() {
+		delegatingresolver.HTTPSProxyFromEnvironment = originalFunc
+	}
+}
+
+// createAndStartBackendServer creates and starts a test backend server,
+// registering a cleanup to stop it.
+func createAndStartBackendServer(t *testing.T) string {
+	backend := &stubserver.StubServer{
+		EmptyCallF: func(context.Context, *testgrpc.Empty) (*testgrpc.Empty, error) { return &testgrpc.Empty{}, nil },
+	}
+	if err := backend.StartServer(); err != nil {
+		t.Fatalf("Failed to start backend: %v", err)
+	}
+	t.Logf("Started TestService backend at: %q", backend.Address)
+	t.Cleanup(backend.Stop)
+	return backend.Address
+}
+
+// setupDNS sets up a manual DNS resolver and registers it, returning the
+// builder for test customization.
+func setupDNS(t *testing.T) *manual.Resolver {
+	manualResolver := manual.NewBuilderWithScheme("dns")
+	originalResolver := resolver.Get("dns")
+	resolver.Register(manualResolver)
+	t.Cleanup(func() {
+		resolver.Register(originalResolver)
+	})
+	return manualResolver
+}
+
+// setupProxy initializes and starts a proxy server, registers a cleanup to
+// stop it, and returns the proxy's listener and helper channels.
+func setupProxy(t *testing.T, backendAddr string, resolutionOnClient bool, reqCheck func(*http.Request) error) (net.Listener, chan error, chan struct{}, chan struct{}) {
+	proxyListener, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		t.Fatalf("failed to listen: %v", err)
+	}
+
+	errCh := make(chan error, 1)
+	doneCh := make(chan struct{})
+	proxyStartedCh := make(chan struct{})
+	t.Cleanup(func() {
+		close(errCh)
+	})
+
+	proxyServer := testutils.NewProxyServer(proxyListener, reqCheck, errCh, doneCh, backendAddr, resolutionOnClient, func() { close(proxyStartedCh) })
+	t.Cleanup(proxyServer.Stop)
+
+	return proxyListener, errCh, doneCh, proxyStartedCh
+}
+
+// TestGrpcDialWithProxy tests grpc.Dial using a proxy and default
+// resolver in the target URI.and verifies that it connects to the proxy server
```

Comment:
typo space and remove period after target URI

Reply:
Done